### PR TITLE
change path of demultiplexing output

### PR DESCRIPTION
**Pull request by karlnyr**
## Description

We currently check for a pattern when doing BCLConvert which does not match the way that we demultiplex. This PR reduces the amount of subdirectories generated in the demultiplexing and changes the output directory of demultiplexing to match the pattern we search for in production today.

The new demultiplexing command requires a specific SampleSheet header to be used, and will throw an error if we use the old one. Code changes to SampleSheet might be needed for this function.

Make sure there are no changes to the bcl2fastq flow.

## Sample sheet

Use SampleSheetCreatorV2 for all novaseq6000 and novaseqx flow cells going forward.

Samplesheet novaseqX section - make software version a constant
`cg demultiplex sample sheet create`

# TO DO

Generate Issue for backwards compatibility for the novaseq6000 flow cells. 

### Changed

- Demultiplexing output directory
- Demultiplexing output structure

### How to prepare for test

- [ ] Ssh to relevant server (depending on type of change)
- [ ] Use stage: `us`
- [ ] Paxa the environment: `paxa`
- [ ] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_main -t cg -b [THIS-BRANCH-NAME] -a
    ```

### How to test

- [ ] Do ...

### Expected test outcome

- [ ] Check that ...
- [ ] Take a screenshot and attach or copy/paste the output.

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/apps/demultiplex/sbatch.py [hunk 1]
```diff
@@ -28,15 +28,13 @@
     "dragen": """
 log "dragen --bcl-conversion-only true \
 --bcl-input-directory {run_dir} \
---output-directory {unaligned_dir} \
---bcl-sampleproject-subdirectories true \
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sbatch.py:32]: @karlnyr This is the change that might cause the current sample sheets to fail right? If we remove the argument we also need to remove the `Sample_Project` column in the sample sheet. Have I understood it correctly?

Comment (karlnyr) [reply cg/apps/demultiplex/sbatch.py:32]: Yes - but by simply using the new SampleSheet -  SampleSheetCreatorV2 - we will have a sample sheet which is compliant with the updated options we use.

File: cg/apps/demultiplex/demultiplex_api.py [hunk 2]
```diff
@@ -227,16 +231,17 @@ def start_demultiplexing(self, flow_cell: FlowCellDirectoryData):
         unaligned_dir: Path = self.get_flow_cell_unaligned_dir(flow_cell=flow_cell)
```
Comment (diitaz93) [inline cg/apps/demultiplex/demultiplex_api.py:231]: There is a method get_flow_cell_unaligned_dir that we could use

File: cg/apps/demultiplex/sample_sheet/index.py [hunk 3]
```diff
@@ -151,7 +151,8 @@ def get_hamming_distance_for_indexes(sequence_1: str, sequence_2: str) -> int:
 
 
 def update_barcode_mismatch_values_for_sample(
-    sample_to_update: FlowCellSampleNovaSeqX, samples_to_compare_to: List[FlowCellSampleNovaSeqX]
+    sample_to_update: FlowCellSampleBCLConvert,
+    samples_to_compare_to: List[FlowCellSampleBCLConvert],
```
Comment (diitaz93) [inline cg/apps/demultiplex/sample_sheet/index.py:155]: These should be `FlowCellSample`

Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/index.py:155]: or union between bcl2fastq and bclconvert sample

File: cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py [hunk 4]
```diff
@@ -136,7 +136,7 @@ def construct_sample_sheet(self) -> List[List[str]]:
 
 
 class SampleSheetCreatorV1(SampleSheetCreator):
-    """Create a raw sample sheet for NovaSeq6000 flow cells."""
+    """Create a raw sample sheet for flow cells."""
```
Comment (diitaz93) [inline cg/apps/demultiplex/sample_sheet/sample_sheet_creator.py:139]:
```suggestion
    """Create a raw sample sheet for flow cells to be demultiplexed with Bcl2fastq."""
```

File: cg/apps/demultiplex/sample_sheet/create.py [hunk 5]
```diff
@@ -22,22 +20,14 @@ def get_sample_sheet_creator(
...
     force: bool,
 ) -> SampleSheetCreator:
-    """Returns an initialised sample sheet creator according to the flow cell sequencer."""
-    sequencer: str = flow_cell.sequencer_type
-    if sequencer == Sequencers.NOVASEQ:
+    """Returns an initialised sample sheet creator according to the software used for demultiplexing."""
+    if bcl_converter == BclConverter.BCL2FASTQ:
         return SampleSheetCreatorV1(
             bcl_converter=bcl_converter, flow_cell=flow_cell, lims_samples=lims_samples, force=force
         )
```
Comment (karlnyr) [inline cg/apps/demultiplex/sample_sheet/create.py:27]: I am not sure if this is correct. I think if we do this change we would also need to change the SampleSheetCreatorV1. Currently, this is a bug in production - we cannot handle BCL2FASTQ sample sheets as they were once generated using the `demultiplexing` repo. We could probably modify SampleSheetCreatorV1 to represent the BCL2FASTQ sample sheets. However, we should probably with that change call it SampleSheetCreatorBcl2Fastq and SampleSheetCreatorBclConvert

File: cg/apps/demultiplex/sample_sheet/models.py [hunk 6]
```diff
@@ -78,8 +63,8 @@ class SampleSheet(BaseModel):
 
 
 class SampleSheetBcl2Fastq(SampleSheet):
-    samples: List[FlowCellSampleNovaSeq6000Bcl2Fastq]
+    samples: List[FlowCellSampleBcl2Fastq]
 
 
 class SampleSheetDragen(SampleSheet):
```
Comment (karlnyr) [inline cg/apps/demultiplex/sample_sheet/models.py:69]: SampleSheetBCLConvert 😢

File: cg/apps/demultiplex/sample_sheet/index.py (hunk 3 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/index.py:155]: nevermind

File: cg/apps/demultiplex/demultiplex_api.py [hunk 7]
```diff
@@ -277,3 +282,12 @@ def start_demultiplexing(self, flow_cell: FlowCellDirectoryData):
...
         return sbatch_number
+
+    @staticmethod
+    def create_demultiplexing_output_dir(
+        flow_cell: FlowCellDirectoryData, demux_dir: Path, unaligned_dir: Path
+    ) -> None:
+        LOG.info(f"Creating demux dir {unaligned_dir}")
+        demux_dir.mkdir(exist_ok=False, parents=True)
+        if flow_cell.bcl_converter == BclConverter.BCL2FASTQ:
+            unaligned_dir.mkdir(exist_ok=False, parents=False)
```
Comment (karlnyr) [inline cg/apps/demultiplex/demultiplex_api.py:293]: Nice! Perhaps remove the LOG.info into a log debug?

File: cg/apps/demultiplex/demultiplex_api.py [hunk 8]
```diff
@@ -73,17 +73,21 @@ def get_sbatch_error(
...
         demux_completed: Path,
         flow_cell: FlowCellDirectoryData,
         environment: Literal["production", "stage"] = "stage",
     ) -> str:
-        """Return sbatch command."""
+        """
+        Return sbatch command.
+        The unaligned_dir is only used for Bcl2Fastq.
+        """
         LOG.info("Creating the sbatch command string")
```
Comment (karlnyr) [inline cg/apps/demultiplex/demultiplex_api.py:86]:
```suggestion
        LOG.debug("Creating the sbatch command string")
```

**Review by karlnyr — COMMENTED** [review]
Neat, I have one major comment though regarding the creation of sample sheets

File: cg/apps/demultiplex/sample_sheet/create.py (hunk 5 above)
Comment (ChrOertlin) [reply cg/apps/demultiplex/sample_sheet/create.py]: fixed

**Comment by ChrOertlin** [comment]
# cg demultiplex samplesheet create 
worked for a novaseq 6000 flow, it created a BCL convert sample sheet

# cg demultiplex samplesheet create -b bcl2fastq
worked for a novaseq 6000 flow cell, a bcl2fastq sample sheet was created

**Comment by ChrOertlin** [comment]
# cg -l DEBUG demultiplex flow-cell -b bcl2fastq
started successfully for a novaseq flow cell
Unaligned directory was created

# cg -l DEBUG demultiplex flow-cell
started sucessfully for a novaseq flow cell 
flat output structure was created (i.e. no Unaligned folder)

**Comment by ChrOertlin** [comment]
Stage:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/christian.oertlin
done.
Log deploy... done.
cg, version 45.4.0

Prod:
Logging deploy ...
Getting deployer... done.
Getting last commit message and SHA... done.
Getting version of deploy scripts... /home/christian.oertlin
done.
Log deploy... done.
cg, version 45.4.0

**Pull request merged**